Makefile.include

```diff
@@ -159,18 +164,21 @@ else
 
 all: link
 
-# Include Board and CPU configuration, if BOARD is not found in BOARDSDIR
-# e.g. when set by the environment fallback to searching in RIOTBOARD
-ifeq (,$(wildcard $(BOARDSDIR)/$(BOARD)/.))
-  ifneq ($(RIOTBOARD),$(BOARDSDIR))
-    # The specified board $(BOARD) was not found in $(BOARDSDIR) fallback to RIOTBOARD
-    BOARDSDIR = $(RIOTBOARD)
-  endif
-  ifeq ($(RIOTBOARD),$(BOARDSDIR))
-    ifeq (,$(wildcard $(BOARDSDIR)/$(BOARD)/.))
-      $(error The specified board $(BOARD) does not exist.)
-    endif
-  endif
+# Folders to search: First the external boards, than the official
+BOARDSDIRS := $(EXTERNAL_BOARD_DIRS) $(RIOTBOARD)
```

Comment:
I don't think `:=` is required here and below. Using just `=` should be enough no ?

Reply:
Just my 2 ¢, but IMHO early evaluation should be preferred unless late evaluation is necessary.
It is faster and leads to less surprising results.

Reply:
I strongly agree with encouraging the use `:=` unless there is a good reason for recursive evaluation. Just recently `make info-boards-supported` started to take 30 seconds on modern notebook hardware instead of less than 1 second, because recursive evaluation can often lead to complexity way above `O(n)`. And switching back to `:=` solved the performance issue. (Some tweaks in addition were added as well, but by far the biggest performance gain was using `:=` over `=`.)

And the typical example of:

```Makefile
a = foo
b = $(b)
a = bar
```

I bet that every C developer would intuitively assume `b` to have the value of `foo`. It just feels unnatural that it now has the value `bar`.

Reply:
> I don't think := is required here and below. Using just = should be enough no ?

> I strongly agree with encouraging the use := unless there is a good reason for recursive evaluation.

Lets not be so black/white here. `:=` is good depending on the case (a summary is present at http://riot-os.org/api/build-system-basics.html). `:=`should be avoided for things that might not be evaluated, our coding conventions state that for this case `memoized` should be used. Using `:=` for everything is also bad, since its always evaluated even when not used. But using `:=` is good for things that do not need to be recursively resolved and where you know it will at least be expanded once. This is the case here `BOARDSDIRS` is **always** expanded, so using `:=` is Ok, and better than `=`, `memoized` would also be detrimental here. 

> It is faster and leads to less surprising results.

This again depends on the case, not faster if never used. `:=` makes order of inclusion mater so might lead to surprising results as well.

> I bet that every C developer would intuitively assume b to have the value of foo. It just feels unnatural that it now has the value bar.

IMO this is an invalid argument, this is not `C` its `make`, don't expect it to behave like `C`, as you don't expect `C` to behave as `python` or `c++`.

Reply:
> our coding conventions state

Not really coding conventions, more of a recommendation.

Reply:
> This is the case here BOARDSDIRS is always expanded, so using := is Ok, and better than =, memoized would also be detrimental here.

Thanks for clarifying this.

Reply:
> > I bet that every C developer would intuitively assume b to have the value of foo. It just feels unnatural that it now has the value bar.
> 
> IMO this is an invalid argument, this is not `C` its `make`, don't expect it to behave like `C`, as you don't expect `C` to behave as `python` or `c++`.

The difference here is that people are using Python or C++ as programming language of their own. No one implements software in Make. People want to implement software in C (at least in the case of RIOT) and have to get around with Make. As a result, developers who spend 99% of the time with C have to understand Makefiles and how also those ~~nasty~~ nice features such as recursive variables work.

To me, having a Makefile that is easy to understand by C developers is an advantage taking the developer base of RIOT into account. Even if this means that the Makefile is less aesthetic from the perspective of a true Make developer (if such people exist).

>>  It is faster and leads to less surprising results.
>
>This again depends on the case, not faster if never used.

This is actually an implementation detail of Make. It would certainly be possible to eliminate dead stores. (I have no idea if Make does perform any optimizations.) But even if it wouldn't, the odds of consistently using `:=` (except where there are reasons to prefer recursive evaluation) is certainly quite good. Even if unused expressions have to be evaluated, that overhead would benefit from the use of  `:=` everywhere else, as only simply evaluated variables are referred to.

So `:=` is potentially more expensive than `=` if the expression is never referred to. It is at least as efficient as `=` when it is referred to once, but often more efficient (as it can only refer to already evaluated variables when `:=` is consistently used). And when the expression is referred to more than once, it is always more efficient.

> := makes order of inclusion mater so might lead to surprising results as well.

First, this is something C developers are used to anyway. Second, it would be IMO less surprising if the order of inclusion does consistently matter, rather than only sporadically. And consistently using recursively expended variables is not only prohibitive expensive performance-wise, it is also not feasible: E.g. the recursive dependency resolution mechanism currently used is not feasible to implement with recursively evaluated variables only.

Reply:
> So := is potentially more expensive than = if the expression is never referred to. It is at least as efficient as = when it is referred to once, but often more efficient (as it can only refer to already evaluated variables when := is consistently used). And when the expression is referred to more than once, it is always more efficient.

That is why `memoized` was added.